File: 2023/23/2323.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
)

type Tile struct {
	i    int16
	n    uint16
	seen []int16
}

func read_file() [][]string {
	const year, day, path string = "2023", "23", "input.txt"
	os.Chdir(year + "/" + day)
	file, err := os.Open(path)
	if err != nil {
		panic("erro ao ler o arquivo")
	}

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanRunes)

	var lines [][]string
	lines = append(lines, []string{})
	line_dx := 0
	for scanner.Scan() {
		text := scanner.Text()
		if text == "\r" {
			continue
		} else if text == "\n" {
			line_dx++
			lines = append(lines, []string{})
			continue
		}
		lines[line_dx] = append(lines[line_dx], text)
	}
	file.Close()
	return lines
}

func main() {
	lines := read_file()
	h := int16(len(lines))
	w := int16(len(lines[0]))

	var done uint16
	paths := []Tile{{0*w + 1, 0, []int16{}}}

	for len(paths) != 0 {

		path := paths[0]
		paths = paths[1:]

		c := path.i / w
		r := path.i - w*c

		// fmt.Println(path.i, r, c)

		if r == h-1 && c == w-2 && path.n > done {
			done = path.n
			continue
		}

		if slices.Contains(path.seen, path.i) {
			continue
		}

		seen := make([]int16, len(path.seen), cap(path.seen))
		copy(seen, path.seen)
		seen = append(seen, path.i)

		for _, dir := range [4][2]int16{{1, 0}, {-1, 0}, {0, 1}, {0, -1}} {
			ndr, ndc := dir[0], dir[1]
			nr := (r) + ndr
			nc := (c) + ndc
			if (nr < 0) || (nr >= h) || (nc < 0) || (nc >= w) {
				continue
			}

			nt := lines[nr][nc]
			if nt == "#" {
				continue
			}
			paths = append(paths, Tile{nc*w + nr, path.n + 1, seen})

		}
	}

	fmt.Println("PARTE 2:", done)
}

File: 2023/23/2323.py
import os
from collections import deque

from icecream import ic

year, day = "2023", "23"
os.chdir(year + "/" + day)
path = open("input.txt", "r").read().strip().split("\n")

# r, c, dr, dc, tiles, seen
tiles = deque([[0, 1, 1, 0, 0, set()]])
done = []
while tiles:
    r, c, dr, dc, n, s = tiles.popleft()
    s = s.copy()
    if (r, c) == (len(path) - 1, len(path[0]) - 2):
        done.append(s)
        continue
    if (r, c) in s:
        continue
    s.add((r, c))
    t = path[r][c]
    if t in "<>^v":
        dirs = ({"<": (0, -1), ">": (0, 1), "^": (-1, 0), "v": (1, 0)}.get(t),)
    else:
        dirs = ((1, 0), (-1, 0), (0, 1), (0, -1))
    for ndr, ndc in dirs:
        nr = r + ndr
        nc = c + ndc
        if nr < 0 or nr >= len(path) or nc < 0 or nc >= len(path[0]):
            continue

        nt = path[nr][nc]
        if nt == "#":
            continue
        tiles.append((nr, nc, ndr, ndc, n + 1, s))

# explored_paths = open("paths.txt", "w")
# for path_done in done:
#     new_path = path.copy()
#     for r, c in path_done:
#         if new_path[r][c] not in "<>^v":
#             new_path[r] = new_path[r][0:c] + "O" + new_path[r][c + 1 :]
#     explored_paths.write(str(len(path_done)) + "\n")
#     for row in new_path:
#         explored_paths.write(row + "\n")

# r, c, dr, dc, tiles, seen

tiles = deque([[0, 1, 1, 0, 0, set()]])
max_path = 0
while tiles:
    r, c, dr, dc, n, s = tiles.popleft()
    s = s.copy()
    if (r, c) == (len(path) - 1, len(path[0]) - 2):
        max_path = max(max_path, n)
        continue
    if (r, c) in s:
        continue
    s.add((r, c))

    for ndr, ndc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
        nr = r + ndr
        nc = c + ndc
        if nr < 0 or nr >= len(path) or nc < 0 or nc >= len(path[0]):
            continue

        nt = path[nr][nc]
        if nt == "#":
            continue
        tiles.append((nr, nc, ndr, ndc, n + 1, s))

print("PARTE 2:", max(map(len, done)))
